radsat-sk/src/tasks/RCommunicationTasks.c

```diff
@@ -0,0 +1,441 @@
+#include <freertos/timers.h>
+
+
+/***************************************************************************************************
+                                   DEFINITIONS & PRIVATE GLOBALS
+***************************************************************************************************/
+
+/** Maximum possible duration of a pass is 15 minutes; value set in ms. */
+#define MAX_PASS_MODE_DURATION	((portTickType)(15*60*1000))
+
+/** Typical duration of quiet mode is 15 minutes; value set in ms. */
+#define QUIET_MODE_DURATION		((portTickType)(15*60*1000))
+
+/** Maximum amount of consecutive NACKs before transmission is aborted. */
+#define NACK_ERROR_LIMIT		((uint8_t)5)
+
+/**
+ * Delay duration (in ms) when the transmitter's buffer is full.
+ * Transmission speed: 9600 bps = 1 byte per ms; This delays long enough to transmit 1 full frame.
+ */
+#define TRANSMITTER_LONG_DELAY	((uint8_t)TRANCEIVER_TX_MAX_FRAME_SIZE)
+
+
+/** Abstraction of the reponse states */
+typedef enum _response_state_t {
+	responseStateIdle	= 0,	///> Awaiting response from Ground Station
+	responseStateReady	= 1,	///> Ready to transmit to Ground Station
+} response_state_t;
+
+
+/** Abstraction of the ACK/NACK return types */
+typedef enum _response_t {
+	responseACK		= 0,	///> Acknowledge (the message was received properly)
+	responseNACK	= 1,	///> Negative Acknowledge (the message was NOT received properly)
+} response_t;
+
+
+/** Abstraction of the communication modes */
+typedef enum _comm_mode_t {
+	commModeQuiet			= -1,	///> Prevent downlink transmissions and automatic state changes
+	commModeIdle			= 0,	///> Not in a pass
+	commModeTelecommand		= 1,	///> Receiving Telecommands from Ground Station
+	commModeFileTransfer	= 2,	///> Transmitting data to the Ground Station
+} comm_mode_t;
+
+
+/** Co-ordinates tasks during the telecommand phase */
+typedef struct _telecommand_state_t {
+	response_state_t transmitReady;	///> Whether the Satellite is ready to transmit a response (ACK, NACK, etc.)
+	response_t responseToSend;		///> What response to send, when ready
+} telecommand_state_t;
+
+
+/** Co-ordinates tasks during the file transfer phase */
+typedef struct _file_transfer_state_t {
+	response_state_t transmitReady;		///> Whether the Satellite is ready to transmit another Frame (telemetry, etc.)
+	response_t responseReceived;		///> What response was received (ACK, NACK, etc.) regarding the previous message
+	uint8_t transmissionErrors;			///> Error counter for recording consecutive NACKs
+} file_transfer_state_t;
+
+
+/** Wrapper structure for communications co-ordination */
+typedef struct _communication_state_t {
+	comm_mode_t mode;					///> The current state of the Communications Tasks
+	telecommand_state_t telecommand;	///> The state during the Telecommand mode
+	file_transfer_state_t fileTransfer;	///> The state during the File Transfer mode
+} communication_state_t;
+
+
+/** Timer for pass mode */
+static xTimerHandle passTimer;
+

```

Comment:
keep in mind that there should be a separate timer for the the beacon

Reply:
Yup, the design was made with that in mind. It won't be included in this PR, but a future one